firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Chunks.cpp

```diff
@@ -0,0 +1,176 @@
+#include "Pufferfish/Protocols/Chunks.h"
+
+#include "Pufferfish/Util/Vector.h"
+#include "catch2/catch.hpp"
+
+namespace PF = Pufferfish;
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(256 bytes)", "[chunks]") {
+  GIVEN("A chunk splitter") {
+    constexpr size_t buffer_size = 256;
+    PF::Protocols::ChunkSplitter<buffer_size, uint8_t> chunks;
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+      PF::Util::Vector<uint8_t, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      auto final_status = chunks.input(val);
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+  }
+}
+
+SCENARIO(
+    "Protocols::ChunkSplitter behaves correctly across chunk boundary(102 bytes)", "[chunks]") {
+  GIVEN("A chunk splitter") {
+    constexpr size_t buffer_size = 102;
+    PF::Protocols::ChunkSplitter<buffer_size, char> chunks;
+
+    WHEN("the input data is within bounds") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+      }
+    }
+
+    WHEN("the input data exceeds bounds without output being called") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+
+      THEN("the final status should be invalid_length") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+
+    WHEN("the input data exceeds bounds after output is called, without delimiter") {
+      uint8_t val = 128;
+      PF::Protocols::ChunkInputStatus status;
+      for (size_t i = 0; i < buffer_size; ++i) {
+        status = chunks.input(val);
+      }
+      PF::Util::Vector<char, buffer_size> buffer;
+      PF::Protocols::ChunkOutputStatus output_status = chunks.output(buffer);
+      auto final_status = chunks.input(val);
+
+      THEN("the final status should be ok") {
+        REQUIRE(status == PF::Protocols::ChunkInputStatus::ok);
+        REQUIRE(output_status == PF::Protocols::ChunkOutputStatus::waiting);
+        REQUIRE(final_status == PF::Protocols::ChunkInputStatus::invalid_length);
+      }
+    }
+
+    WHEN("output is called on an input of invalid length") {
```

Comment:
Is this equivalent to "WHEN more data is input than can be held by the splitter's buffer: THEN calling the output method returns invalid_length"?

Reply:
Yes! added these two cases in different given as suggested.